module/synchronization/core.go

```diff
@@ -287,8 +344,9 @@ func (c *Core) getRequestableItems() ([]uint64, []flow.Identifier) {
 			continue
 		}
 
-		// if we reached maximum number of attempts, delete
-		if status.Attempts >= c.Config.MaxAttempts {
+		// if we reached maximum number of attempts and we're not certain that the height
+		// has actually been finalized, delete
+		if height > c.maxHeightReceived && status.Attempts >= c.Config.MaxAttempts {
 			delete(c.heights, height)
 			continue
 		}
```

Comment:
> Modify the status tracking logic so that we only delete requests that have reached the max number of attempts if we haven't already received a finalized block with higher height

I think the reason for this change is to make sure we don't _stop requesting a block height_ by deleting it. But when we delete a height status like this, if we haven't received the block yet, practically it will just reset the request retry timer (the height will be re-queued [here](https://github.com/onflow/flow-go/blob/982a0a2f01d2aff611d19c353f2b361408de6476/module/synchronization/core.go#L98-L112)). So I'm not sure there's any reason to keep track of `maxHeightReceived`. We could add a note about deleted height statuses to the docs though.

Reply:
That's correct, but that assumes that we will eventually receive another sync response from another node which triggers the `HandleHeight` at a higher height. In practice this is probably a safe assumption, so I don't mind removing this.

Reply:
If we don't receive sync responses from other nodes, we never send range requests at all and the engine doesn't really work, so I think it's a safe assumption.